### DISCUSSION TITLE
Having trouble dumping saves of any Flash and SRAM save type N64 games

### DISCUSSION THREAD

**Discussion by KayodeBakker** [sanni/cartreader#1011] [Need help]
Hello,

I recently bought a V3 reader from save the hero builders and I've been in the process of dumping all my old Nintendo cartridges and I'm stuck on a specific few of the N64 cartridges. All problems seem to be with N64 games that use SRAM or Flash for save game storage. The games specifically are:
- Super Smash Bros.
- Pokemon Stadium
- Pokémon Stadium 2
- Pokémon Snap
- The Legend of Zelda - Ocarina of Time
- The Legend of Zelda - Majora's Mask

The reader I bought has the SNES SA-1 unlock & N64 EEPROM save + RTC options included which seems to work as I have managed to dump other N64 games including EEPROM saves. The games mentioned above are the only games where the saves not what I expect. The roms, however dump just fine.
What do I mean with 'what I expect'? Well the reader doesn't give me any errors and seem to create save files on the MicroSD card except when trying to use these saves via emulators such as Mupen64++ or Project64, I notice that the saves are considered to be freshly made as no data is to be found.

As for the switch settings:
- Volt set to 3V
- EEP set to right (on)
- CLK0 set to left (off)
- CLK1 set to right (on)

I'm unsure what is happening and especially since the game roms are being dumped and are usable AND other games (such as EEPROM type ones) have functional saves.

I first asked this question at the retailer but they mentioned that since the roms are being dumped just fine, that I should check here.
Can someone help me with this by any chance?

**Comment by sanni** [comment]
Have you tried this? https://github.com/sanni/cartreader/discussions/593

As an alternative you can also use any hex editor. Just create a savegame with the emulator open it up in the hex editor and compare it to the save you dumped from the cartridge, then byte/word swap the cart save around until it matches the emu save.

This could also work: https://savefileconverter.com/#/flash-carts

**Comment by KayodeBakker** [comment]
I have tried #593 and also the link you posted but neither seem to do the trick.
I honestly don't know where to start with a hex editor so I'm unsure what to do.

**Comment by KayodeBakker** [comment]
@sanni 
I also tried using this https://game-tools-collection.com/super-smash-bros-n64/save-editor and it thinks that the dumped save isn't valid.

**Comment by sanni** [comment]
Can you put your saves into a zip file and upload them here? Then I convert them for emulator and document it for the Wiki.

**Comment by KayodeBakker** [comment]
[N64 saves SRAM+FLASH.zip](https://github.com/user-attachments/files/16551374/N64.saves.SRAM%2BFLASH.zip)
I've added Pokémon 2, Pokémon Snap, Ocarina of Time and Super Smash Bros. in a zip. After your first comment I went back and did all the saves via those tools you mentioned and only Ocarina of Time and Majora's Mask seem to work after converting. The rest however didn't change. The saves in this zip is exactly as the reader I have delivered it.

**Comment by KayodeBakker** [comment]
I'd like to add that I noticed that save the hero builders supplies this reader with a V13.1 sanni release. Would upgrading matter in any way?

**Comment by KayodeBakker** [comment]
@sanni No rush of course but do you have any inkling what's the issue?

**Reply by sanni** [reply]
Seem like bad dumps to me. Maybe try cleaning the cartridges with q-tip and rubbing alcohol.
Since Ocarina of Time is SRAM and Majoras Mask is Flashram, all save types seem to dump correctly on your Reader, generally speaking. 

What region are the games? I could try to load the saves with my Everdrive then we would know for sure if they are bad dumps.

**Reply by KayodeBakker** [reply]
Actually already did that beforehand but I'll try again. Especially as these games boot up just fine on the console itself I find it weird if that would be the problem.

All of these are PAL region cartridges.

**Reply by sanni** [reply]
The Zelda save worked on my flash cart, the rest didn't

**Reply by SAVE-THE-HERO** [reply]
Until now, There have been no instances where a hardware glitch has prevented the backup of save data for a particular game.
@KayodeBakker  If the discussion here does not solve the problem, please email me again.
I hope this issue will be resolved.

**Reply by KayodeBakker** [reply]
@sanni 
@SAVE-THE-HERO 
I will try to clean all carts more intensely just in case. I did clean them until the q tips weren't black anymore but maybe something else made them dirty again.

I'll get back to you soon.

**Reply by sanni** [reply]
I found a cart where I could not consistently dump the SRAM save correctly. I'll see if I can improve the dumping code. One thing I noticed, it dumped perfectly if I played it in the N64 just before dumping the save.

**Reply by KayodeBakker** [reply]
Interesting! I'm unsure if I tried that so I'll try that today.

**Reply by KayodeBakker** [reply]
I'm not sure I understand the logic behind it but before recleaning the contacts once again I tried @sanni 's suggestion and that actually works. I place either a Flash or SRAM cartridge in the N64 and after it boots I immediately turn the N64 off and place the cartridge in the reader. When I dump the save right then and there it would still need a conversion but afterwards it works just fine.

Great that it works but can someone explain this to me?

**Reply by KayodeBakker** [reply]
@sanni 
So I can close the discussion here, could you explain the logic behind this solution for me?

**Reply by sanni** [reply]
After changing the OSCRs read timings a bit the SRAM on my cartridge now dumps consistently even without tricks. 
So I would guess the previous timings had a high chance to "confuse" the SRAM which resulted in bad dumps. Playing the cart in a real console probably just "reset" everything back to normal. Waiting until all the capacitors inside the cartridge have drained would probably have the same effect of "resetting" it.

**Reply by KayodeBakker** [reply]
Interesting! Is there something I need to update in the reader?

**Reply by sanni** [reply]
I have put together a new release [here](https://github.com/sanni/cartreader/releases/tag/V14.3).
If you want you can update to that and see if this fixes reading SRAM/Flashram saves for you. If not then I will have to go over the code again. 😄

**Reply by KayodeBakker** [reply]
I'll have to see how to do that, but I will try that today or tomorrow!
